sycl/plugins/Intel_level0/pi_level0.cpp

```diff
@@ -0,0 +1,3723 @@
+    zePrint("piEnqueueMemBufferMap: duplicate mapping detected\n");
+    return PI_INVALID_OPERATION;
+  } else {
+    Mappings.insert({MappedTo, {Offset, Size}});
+  }
+  return PI_SUCCESS;
+}
+
+pi_result _pi_mem::removeMapping(void *MappedTo, Mapping &MapInfo) {
+  std::lock_guard<std::mutex> Lock(MappingsMutex);
+  auto It = Mappings.find(MappedTo);
+  if (It == Mappings.end()) {
+    zePrint("piEnqueueMemUnmap: unknown memory mapping\n");
+    return PI_INVALID_VALUE;
+  }
+  MapInfo = It->second;
+  Mappings.erase(It);
+  return PI_SUCCESS;
+}
+
+ze_result_t
+_pi_context::getFreeSlotInExistingOrNewPool(ze_event_pool_handle_t &ZePool,
+                                            size_t &Index) {
+  // Maximum number of events that can be present in an event ZePool is captured
+  // here Setting it to 256 gave best possible performance for several
+  // benchmarks
+  static const char *MaxNumEventsPerPoolEnv =
+      std::getenv("MAX_NUMBER_OF_EVENTS_PER_EVENT_POOL");
+  static const pi_uint32 MaxNumEventsPerPool =
+      (MaxNumEventsPerPoolEnv) ? std::atoi(MaxNumEventsPerPoolEnv) : 256;
+
+  Index = 0;
+  // Create one event ZePool per MaxNumEventsPerPool events
+  if ((ZeEventPool == nullptr) ||
+      (NumEventsAvailableInEventPool[ZeEventPool] == 0)) {
+    // Creation of the new ZePool with record in NumEventsAvailableInEventPool
+    // and initialization of the record in NumEventsLiveInEventPool must be done
+    // atomically. Otherwise it is possible that decrementAliveEventsInPool will
+    // be called for the record in NumEventsLiveInEventPool before its
+    // initialization.
+    std::lock(NumEventsAvailableInEventPoolMutex,
+              NumEventsLiveInEventPoolMutex);
+    std::lock_guard<std::mutex> NumEventsAvailableInEventPoolGuard(
+        NumEventsAvailableInEventPoolMutex, std::adopt_lock);
+    std::lock_guard<std::mutex> NumEventsLiveInEventPoolGuard(
+        NumEventsLiveInEventPoolMutex, std::adopt_lock);
+
+    ze_event_pool_desc_t ZeEventPoolDesc;
+    ZeEventPoolDesc.count = MaxNumEventsPerPool;
+    ZeEventPoolDesc.flags = ZE_EVENT_POOL_FLAG_TIMESTAMP;
+    ZeEventPoolDesc.version = ZE_EVENT_POOL_DESC_VERSION_CURRENT;
+
+    ze_device_handle_t ZeDevice = Device->ZeDevice;
+    if (ze_result_t ZeRes =
+            zeEventPoolCreate(Device->Platform->ZeDriver, &ZeEventPoolDesc, 1,
+                              &ZeDevice, &ZeEventPool))
+      return ZeRes;
+    NumEventsAvailableInEventPool[ZeEventPool] = MaxNumEventsPerPool - 1;
+    NumEventsLiveInEventPool[ZeEventPool] = MaxNumEventsPerPool;
+  } else {
+    std::lock_guard<std::mutex> NumEventsAvailableInEventPoolGuard(
+        NumEventsAvailableInEventPoolMutex);
+    Index = MaxNumEventsPerPool - NumEventsAvailableInEventPool[ZeEventPool];
+    --NumEventsAvailableInEventPool[ZeEventPool];
+  }
+  ZePool = ZeEventPool;
+  return ZE_RESULT_SUCCESS;
+}
+
+ze_result_t
+_pi_context::decrementAliveEventsInPool(ze_event_pool_handle_t ZePool) {
```

Comment:
If number of live events oscillates around MaxNumEventsPerPool, pool creation/destruction will be called frequently. I'm not sure how practical this scenario is, maybe it is worth adding a note or TODO.

Reply:
Sorry, I am not sure this case is possible with the current implementation. Number of live events is only decremented starting from MaxNumEventsPerPool, it is not incremented, so number of live events cannot oscillate around some number.